osu.Game/Screens/Select/PlaySongSelect.cs

```diff
@@ -92,6 +92,13 @@ protected override bool OnStart()
         {
             if (playerLoader != null) return false;
 
+            var validMods = Mods.Value.ToArray();
+
+            if (!ModUtils.CheckModsToRemove(validMods, out var removed, Beatmap.Value.BeatmapInfo.Difficulty))
+                validMods = validMods.Except(removed).ToArray();
```

Comment:
Without testing I have 95% confidence that placing this logic in `PlaySongSelect` is not the right place. Rationale being:

- it will not work when selecting beatmaps to add to the playlist in multiplayer
- it will not work when selecting free mods in multiplayer
- it will not work when selecting beatmaps to add to the playlist in playlists

I'm not yet sure we want to introduce this complexity to the client, as it complicates the already-way-too-complex mod logic further, but if we do, then it better work everywhere.